src/engine.rs

```diff
@@ -185,7 +185,7 @@ impl Module {
         }
     }
 
-    fn outputs(&self) -> &[LineType] {
+    fn outputs(&self) -> Vec<Terminal> {
```

Comment:
I'd like to keep these return values as references borrowed from self. `inputs` and `outputs` will be called every tick so we need to be careful about allocations

Reply:
I don't think it let me... This was my reaction to the borrow checker. It complained about a reference to temporary data inside the function, but I'll switch it back and see if there's another way. Will reach out if I'm stuck.

Reply:
If you stick these vecs on the struct itself and initialise them in `create`, you should be able to just return refs to them in `inputs` and `outputs`

Reply:
Ah I thought that this _might_ have been what you proposed in WhatsApp. I'll do that.